### PR TITLE
[SPARK-8420][SQL] Fix comparision of timestamps/dates with strings

### PR DESCRIPTION
**Pull request by marmbrus**
In earlier versions of Spark SQL we casted `TimestampType` and `DataType` to `StringType` when it was involved in a binary comparison with a `StringType`.  This allowed comparing a timestamp with a partial date as a user would expect.
- `time > "2014-06-10"`
- `time > "2014"`

In 1.4.0 we tried to cast the String instead into a Timestamp.  However, since partial dates are not a valid complete timestamp this results in `null` which results in the tuple being filtered.

This PR restores the earlier behavior.  Note that we still special case equality so that these comparisons are not affected by not printing zeros for subsecond precision.

### CODE REVIEW

**Comment by marmbrus** [comment]
@yhuai updated to avoid changing equality behavior.

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameDateSuite.scala [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+      (1, Date.valueOf("2015-01-01")),
+      (2, Date.valueOf("2014-01-01"))).toDF("i", "t")
+
+    checkAnswer(
+      df.select("t").filter($"t" <= "2014-06-01"),
+      Row(Date.valueOf("2014-01-01")) :: Nil)
+
+
+    checkAnswer(
+      df.select("t").filter($"t" >= "2014-06-01"),
```
Comment (yhuai) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameDateSuite.scala]: Very minor: change it to `$"t" >= "2014"`?

**Comment by yhuai** [comment]
lgtm

**Comment by yhuai** [comment]
Thanks! I am merging it to master and branch 1.4.

**Comment by yhuai** [comment]
It does not merge cleanly with branch 1.4. Will have another pr for 1.4 backport.

**Pull request closed without merging**